File: frontend/src/api/societies.ts
import { apiClient } from "@/api/client";
import type { PropertyOut } from "@/api/properties";
import type { HouseType, LocationType, SocietyStatus } from "@/types/enums";

export interface SocietyOut {
  id: string;
  name: string;
  code: string;
  status: SocietyStatus;
  address: string | null;
  city: string | null;
  state: string | null;
  pincode: string | null;
  latitude: number | null;
  longitude: number | null;
  created_at: string;
}

export interface SocietyLocationOut {
  id: string;
  society_id: string;
  name: string;
  location_type: LocationType;
  created_at: string;
}

export interface SocietyLookupOut {
  id: string;
  name: string;
}

export interface SocietySearchResultOut {
  id: string;
  name: string;
  city: string | null;
}

export interface SocietyCreateInput {
  name: string;
  address: string;
  city: string;
  state: string;
  pincode: string;
  /** Optional, along with latitude/longitude below — a Platform Owner
   * can add Wings/Rows right at creation, or leave this empty and let
   * the Admin add them later from the Properties page. `code` is never
   * sent — the backend auto-generates a guaranteed-unique one. */
  locations?: { name: string; location_type: LocationType }[];
  /** GPS pin — optional, but both-or-neither (backend rejects a lone
   * coordinate). */
  latitude?: number;
  longitude?: number;
}

export interface SocietyUpdateInput {
  name: string;
  address: string;
  city: string;
  state: string;
  pincode: string;
  latitude?: number;
  longitude?: number;
}

export const societiesApi = {
  /** Platform Owner creates a society directly from their dashboard —
   * the only way a society comes into existence now. ACTIVE immediately. */
  create: (input: SocietyCreateInput) => apiClient.post<SocietyOut>("/societies", input),
  /** Platform Owner edits name/address/city/state/pincode after creation
   * — `code` stays fixed (see backend SocietyUpdateIn's docstring). */
  update: (id: string, input: SocietyUpdateInput) => apiClient.patch<SocietyOut>(`/societies/${id}`, input),
  /** Public — used by the Admin/Resident signup forms to resolve a
   * society code to its id/name before submitting. */
  lookup: (code: string) => apiClient.get<SocietyLookupOut>(`/societies/lookup/${encodeURIComponent(code)}`),
  /** Public — name-search picker alternative to typing the exact code.
   * Backend enforces a 3-char minimum and rate-limits by IP; never
   * returns `code` (see SocietySearchResultOut). */
  search: (q: string) => apiClient.get<SocietySearchResultOut[]>("/societies/search", { params: { q } }),
  list: () => apiClient.get<SocietyOut[]>("/societies"),
  approve: (id: string) => apiClient.post<SocietyOut>(`/societies/${id}/approve`),
  updateStatus: (id: string, status: SocietyStatus) =>
    apiClient.patch<SocietyOut>(`/societies/${id}/status`, { status }),
  /** Platform Owner viewing/adding a society's Wings/Rows straight from
   * the Societies page's Edit modal — same data the Admin manages from
   * their own Properties page, reachable without switching roles. */
  listLocations: (id: string) => apiClient.get<SocietyLocationOut[]>(`/societies/${id}/locations`),
  addLocation: (id: string, name: string, locationType: LocationType) =>
    apiClient.post<SocietyLocationOut>(`/societies/${id}/locations`, { name, location_type: locationType }),
  /** Renaming is always allowed; changing WING<->ROW only succeeds while
   * no Property yet points at it (backend rejects with a 409 naming how
   * many properties are in the way). */
  updateLocation: (societyId: string, locationId: string, name: string, locationType: LocationType) =>
    apiClient.patch<SocietyLocationOut>(`/societies/${societyId}/locations/${locationId}`, {
      name, location_type: locationType,
    }),
  /** Bulk-generates every tower/floor/flat combination in one request —
   * one Wing per entry in `wings`, each with its own floor count and
   * flats/floor (a taller tower next to a shorter one is fine). A Wing
   * with no name gets an auto-generated one server-side ("Tower N").
   * `floor_overrides` lists per-floor exceptions to that Wing's default
   * flats_per_floor (e.g. a ground floor with fewer flats). */
  generateFlatsStructure: (
    id: string,
    wings: {
      name?: string;
      floor_count: number;
      flats_per_floor: number;
      floor_overrides?: { floor_number: number; flats: number }[];
    }[]
  ) => apiClient.post<PropertyOut[]>(`/societies/${id}/structure/flats`, { wings }),
  /** Bulk-generates every row/house at ground-floor-only — extra storeys
   * per house are set afterward via updatePropertyFloors. */
  generateBungalowStructure: (id: string, rowCount: number, housesPerRow: number) =>
    apiClient.post<PropertyOut[]>(`/societies/${id}/structure/bungalows`, {
      row_count: rowCount, houses_per_row: housesPerRow,
    }),
  listProperties: (id: string) => apiClient.get<PropertyOut[]>(`/societies/${id}/properties`),
  updatePropertyFloors: (societyId: string, propertyId: string, floorsAboveGround: number) =>
    apiClient.patch<PropertyOut>(`/societies/${societyId}/properties/${propertyId}/floors`, {
      floors_above_ground: floorsAboveGround,
    }),
  /** The Society Mapping page's Flats & Houses-mapping step — one Property
   * with a specific, hand-typed house_number on a chosen Wing + floor
   * (FLAT) or Row (BUNGALOW, no floor). */
  addProperty: (
    societyId: string, locationId: string, houseNumber: string, houseType: HouseType, floorNumber?: number
  ) =>
    apiClient.post<PropertyOut>(`/societies/${societyId}/properties`, {
      location_id: locationId, house_number: houseNumber, house_type: houseType, floor_number: floorNumber,
    }),
  /** Correcting a Wing/Row/floor/house-number typo made while mapping —
   * full replace, same shape as addProperty. */
  updateProperty: (
    societyId: string, propertyId: string, locationId: string, houseNumber: string, houseType: HouseType,
    floorNumber?: number
  ) =>
    apiClient.patch<PropertyOut>(`/societies/${societyId}/properties/${propertyId}`, {
      location_id: locationId, house_number: houseNumber, house_type: houseType, floor_number: floorNumber,
    }),
  /** Removing a wrongly-added Property — only succeeds while nothing else
   * (a Resident, a payment...) references it yet. */
  deleteProperty: (societyId: string, propertyId: string) =>
    apiClient.delete(`/societies/${societyId}/properties/${propertyId}`),
};

export const locationsApi = {
  list: () => apiClient.get<SocietyLocationOut[]>("/properties/locations"),
  create: (name: string, locationType: LocationType) =>
    apiClient.post<SocietyLocationOut>("/properties/locations", { name, location_type: locationType }),
};

export const propertyAdminApi = {
  create: (locationId: string, houseNumber: string, houseType: HouseType, floorNumber?: number) =>
    apiClient.post("/properties", { location_id: locationId, house_number: houseNumber, house_type: houseType, floor_number: floorNumber }),
};

File: backend/app/api/v1/societies.py
"""Societies endpoints — public signup + Platform Owner administration."""
import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Request, status

from app.core.db import get_db
from app.core.security import CurrentUser, require_role
from app.models.enums import Role
from app.schemas.property import (
    BungalowStructureIn,
    FlatsStructureIn,
    PropertyCreateIn,
    PropertyFloorsUpdateIn,
    PropertyOut,
    SocietyLocationCreateIn,
    SocietyLocationOut,
    SocietyLocationUpdateIn,
)
from app.schemas.society import (
    SocietyCreateIn,
    SocietyLookupOut,
    SocietyOut,
    SocietySearchResultOut,
    SocietySignupIn,
    SocietySignupOut,
    SocietyStatusUpdateIn,
    SocietyUpdateIn,
)
from app.services import property_service, society_service
from sqlalchemy.ext.asyncio import AsyncSession

router = APIRouter(prefix="/societies", tags=["societies"])


@router.get("/lookup/{code}", response_model=SocietyLookupOut)
async def lookup(code: str, db: Annotated[AsyncSession, Depends(get_db)]) -> SocietyLookupOut:
    """Public — used by the Admin/Resident signup forms to find their
    society by its code without needing its internal UUID, and without
    exposing the full society list."""
    society = await society_service.lookup_society_by_code(db, code)
    if society is None:
        raise HTTPException(status.HTTP_404_NOT_FOUND, "Society not found")
    return SocietyLookupOut(id=society.id, name=society.name)


@router.get("/search", response_model=list[SocietySearchResultOut])
async def search(
    q: str,
    request: Request,
    db: Annotated[AsyncSession, Depends(get_db)],
) -> list[SocietySearchResultOut]:
    """Public — lets the Admin/Resident signup forms offer a name-search
    picker as an alternative to typing the exact code. Narrower than
    `lookup` above only in that it matches by substring instead of an
    exact code; see society_service.search_societies_by_name for the
    limits (min query length, capped results, per-IP rate limit) that
    keep this from becoming a full-directory-enumeration endpoint."""
    client_ip = request.client.host if request.client else "unknown"
    societies = await society_service.search_societies_by_name(db, q, client_ip)
    return [SocietySearchResultOut(id=s.id, name=s.name, city=s.city) for s in societies]


@router.post("/signup", response_model=SocietySignupOut)
async def signup(
    body: SocietySignupIn, db: Annotated[AsyncSession, Depends(get_db)]
) -> SocietySignupOut:
    """Public — no auth required. Creates a PENDING society + PENDING Admin
    together; both wait for Platform Owner approval (Section 26)."""
    society, admin = await society_service.signup_society_and_admin(db, body)
    return SocietySignupOut(society_id=society.id, admin_user_id=admin.id)


@router.get("", response_model=list[SocietyOut])
async def list_all(
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.PLATFORM_OWNER))],
) -> list[SocietyOut]:
    societies = await society_service.list_societies(db)
    return [SocietyOut.model_validate(s) for s in societies]


@router.post("", response_model=SocietyOut)
async def create(
    body: SocietyCreateIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.PLATFORM_OWNER))],
) -> SocietyOut:
    """A society now only comes into existence this way — created directly
    by the Platform Owner from their dashboard, ACTIVE immediately. Admin
    signup (POST /admins/signup) targets an existing society created here;
    it no longer creates one itself."""
    society = await society_service.create_society(db, body)
    return SocietyOut.model_validate(society)


@router.post("/{society_id}/approve", response_model=SocietyOut)
async def approve(
    society_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.PLATFORM_OWNER))],
) -> SocietyOut:
    society = await society_service.approve_society_and_admin(db, society_id, current.user_id)
    return SocietyOut.model_validate(society)


@router.patch("/{society_id}", response_model=SocietyOut)
async def update_profile(
    society_id: uuid.UUID,
    body: SocietyUpdateIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.PLATFORM_OWNER))],
) -> SocietyOut:
    """Edits name/address/city/state/pincode — `code` and locations aren't
    editable here (see SocietyUpdateIn's docstring)."""
    society = await society_service.update_society_profile(db, society_id, body)
    return SocietyOut.model_validate(society)


@router.get("/{society_id}/locations", response_model=list[SocietyLocationOut])
async def list_locations(
    society_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.PLATFORM_OWNER))],
) -> list[SocietyLocationOut]:
    """Platform Owner viewing a society's Wings/Rows from the Societies
    page's Edit modal, without switching into that society's Admin role."""
    locations = await society_service.list_society_locations(db, society_id)
    return [SocietyLocationOut.model_validate(loc) for loc in locations]


@router.post("/{society_id}/locations", response_model=SocietyLocationOut)
async def add_location(
    society_id: uuid.UUID,
    body: SocietyLocationCreateIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.PLATFORM_OWNER))],
) -> SocietyLocationOut:
    """Platform Owner adding a Wing/Row to an existing society — same
    table the Admin's own POST /properties/locations writes to."""
    location = await society_service.add_society_location(db, society_id, body)
    return SocietyLocationOut.model_validate(location)


@router.patch("/{society_id}/locations/{location_id}", response_model=SocietyLocationOut)
async def update_location(
    society_id: uuid.UUID,
    location_id: uuid.UUID,
    body: SocietyLocationUpdateIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.PLATFORM_OWNER))],
) -> SocietyLocationOut:
    """Platform Owner renaming (or, if unused, retyping) an existing
    Wing/Row from the Edit modal."""
    location = await society_service.update_society_location(db, society_id, location_id, body)
    return SocietyLocationOut.model_validate(location)


@router.post("/{society_id}/structure/flats", response_model=list[PropertyOut])
async def generate_flats_structure(
    society_id: uuid.UUID,
    body: FlatsStructureIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.PLATFORM_OWNER))],
) -> list[PropertyOut]:
    """Bulk-generates this society's Towers + Flats in one shot — see
    FlatsStructureIn's docstring."""
    properties = await society_service.generate_flats_structure(db, society_id, body)
    return [PropertyOut.model_validate(p) for p in properties]


@router.post("/{society_id}/structure/bungalows", response_model=list[PropertyOut])
async def generate_bungalow_structure(
    society_id: uuid.UUID,
    body: BungalowStructureIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.PLATFORM_OWNER))],
) -> list[PropertyOut]:
    """Bulk-generates this society's Rows + Houses in one shot — see
    BungalowStructureIn's docstring."""
    properties = await society_service.generate_bungalow_structure(db, society_id, body)
    return [PropertyOut.model_validate(p) for p in properties]


@router.get("/{society_id}/properties", response_model=list[PropertyOut])
async def list_society_properties(
    society_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.PLATFORM_OWNER))],
) -> list[PropertyOut]:
    """Platform Owner viewing a society's generated houses — mainly so the
    Edit modal can offer the per-house floors_above_ground follow-up step
    after a bulk BUNGALOW generation, and so the Society Mapping page can
    show what's already on record."""
    properties = await property_service.list_properties(db, society_id)
    return [PropertyOut.model_validate(p) for p in properties]


@router.post("/{society_id}/properties", response_model=PropertyOut)
async def add_society_property(
    society_id: uuid.UUID,
    body: PropertyCreateIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.PLATFORM_OWNER))],
) -> PropertyOut:
    """Platform Owner adding one Property with a specific, hand-typed
    house_number — the Society Mapping page's Flats-mapping step, for a
    Wing/floor combination where the numbering isn't the simple
    sequential scheme the bulk generator assumes."""
    prop = await society_service.add_society_property(db, society_id, body)
    return PropertyOut.model_validate(prop)


@router.patch("/{society_id}/properties/{property_id}/floors", response_model=PropertyOut)
async def update_property_floors(
    society_id: uuid.UUID,
    property_id: uuid.UUID,
    body: PropertyFloorsUpdateIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.PLATFORM_OWNER))],
) -> PropertyOut:
    prop = await society_service.update_property_floors_above_ground(
        db, society_id, property_id, body.floors_above_ground
    )
    return PropertyOut.model_validate(prop)


@router.patch("/{society_id}/status", response_model=SocietyOut)
async def update_status(
    society_id: uuid.UUID,
    body: SocietyStatusUpdateIn,
    db: Annotated[AsyncSession, Depends(get_db)],
    current: Annotated[CurrentUser, Depends(require_role(Role.PLATFORM_OWNER))],
) -> SocietyOut:
    society = await society_service.update_society_status(db, society_id, body.status)
    return SocietyOut.model_validate(society)
